cockatrice/src/dlg_connect.cpp

```diff
@@ -156,26 +167,104 @@ void DlgConnect::actSaveConfig()
     }
 
-    rebuildComboBoxList();
+    preRebuildComboBoxList();
 }
 
-void DlgConnect::rebuildComboBoxList()
+void DlgConnect::downloadPublicServers()
 {
     previousHosts->clear();
+    previousHosts->addItem(placeHolderText);
+
+    QUrl url(QString(PUBLIC_SERVERS_JSON));
+    auto *nam = new QNetworkAccessManager(this);
+    QNetworkReply *reply = nam->get(QNetworkRequest(url));
+    connect(reply, SIGNAL(finished()), this, SLOT(actFinishParsingDownloadedData()));
+}
+
+void DlgConnect::actFinishParsingDownloadedData()
+{
+    auto *reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    if (errorCode == QNetworkReply::NoError) {
+        // Get current saved hosts
+        UserConnection_Information uci;
+        savedHostList = uci.getServerInfo();
+
+        // List of public servers
+        // Will be populated below
+        QStringList publicServersAvailable;
+        QStringList publicServersToRemove;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
+        for (const auto &server : publicServersJSONList) {
+            // Data inside one server at a time
+            // server: [{ ... }, ..., { ... }]
+            const auto serverMap = server.toMap();
+
+            QString serverName = serverMap["name"].toString();
+
+            if (serverMap["isInactive"].toBool()) {
+                publicServersToRemove.append(serverName);
+                continue;
+            }
 
+            QString serverAddress = serverMap["host"].toString();
+            QString serverPort = serverMap["port"].toString();
+
+            publicServersAvailable.append(serverName);
+
+            if (!savedHostList.contains(serverName)) {
+                settingsCache->servers().addNewServer(serverName, serverAddress, serverPort, "", "", false);
+            } else {
+                settingsCache->servers().updateExistingServerWithoutLoss(serverName, serverAddress, serverPort);
+            }
+        }
+
+        // If a server was removed from the public list,
+        // we will delete it from the local system.
+        // Will not delete "unofficial" servers
+        for (auto server : savedHostList) {
+            QString serverName = server.getSaveName();
+            bool isCustom = server.isCustomServer();
+
+            if (!isCustom && publicServersToRemove.indexOf(serverName) != -1) {
+                settingsCache->servers().removeServer(serverName);
+            }
+        }
+
+        reply->deleteLater();
+    } else {
+        qDebug() << "[CONNECTION DIALOG]"
+                 << "Error Downloading Public Servers" << errorCode;
+    }
+
+    // Signal that we're done with updating the combo box
+    emit sigPublicServersDownloaded();
```

Comment:
What do you think about having separate signals for success/fail? Or if one signal, at least include a data object that gives info about the failure. I don't like swallowing it.

Reply:
The signal's job here is to just say "Hey we're done, go rebuild" It has no other meaning. If you want a 2nd signal possible to be emitted and all of them caught so be it